clj-refactor.el

```diff
@@ -1957,35 +1957,98 @@ following this convention: https://stuartsierra.com/2015/05/10/clojure-namespace
+                               (symbol-name alias-name)
+                               (string-join (seq-map #'symbol-name lang-contexts) ","))
+                     (format "[%s :as %s]"
+                             (symbol-name alias-require)
+                             (symbol-name alias-name)))
+                   (list alias-name alias-require)
+                   alias-index)))
+      (let ((ns-require (completing-read "Add require: " alias-index)))
+        (gethash ns-require alias-index))))
+   ((= (length candidates) 1)
+     (cl-first candidates))))
 
 (defun cljr--js-alias-p (alias)
   (and (cljr--cljs-file-p)
        (string-equal "js" alias)))
 
+(defun cljr--ns-short-alias-at-point ()
+  "Returns the (alias)/ just prior to the point excluding the trailing slash."
+  (thread-last (buffer-substring-no-properties
+                (cljr--point-after 'paredit-backward)
+                (1- (point)))
+               (string-remove-prefix "::")
+               (string-remove-prefix "@")))
+
 (defun cljr--magic-requires-lookup-alias ()
-  "Return (alias (ns.candidate1 ns.candidate1)) if we recognize
-the alias in the project."
-  (let ((short (thread-last (buffer-substring-no-properties
-                             (cljr--point-after 'paredit-backward)
-                             (1- (point)))
-                 (string-remove-prefix "::")
-                 (string-remove-prefix "@"))))
+  "Generate candidates requires based on the alias at point.
+
+  Constructs a list of alias, namespace, language-contexts
+  triplets from either the middleware or `cljr-magic-require-namespaces'.
+
+  Preference for namespaces candidates with an exact match on
+  alias, namespace and then a fallback to use
+  `cljr-magic-require-namespaces'."
+  (let ((short (cljr--ns-short-alias-at-point)))
     (unless (or (cljr--resolve-alias short)
                 (cljr--js-alias-p short))
-      (if-let ((aliases (ignore-errors (cljr--get-aliases-from-middleware)))
-               (candidates (gethash (intern short) aliases)))
-          (list short candidates)
+      (if-let ((candidates
+                (seq-filter (lambda (elt)
+                              ;; prefer to match alias
+                              (or (equal (intern short) (car elt))
+                                  ;; but fallback on matching full namespace
+                                  (equal short (symbol-name (cadr elt)))))
+                            (cljr--namespace-aliases))))
+          candidates
         (when (and cljr-magic-require-namespaces ; a regex against "" always triggers
                    (string-match-p (cljr--magic-requires-re) short))
           ;; This when-let might seem unnecessary but the regexp match
           ;; isn't perfect.
           (when-let (long (cljr--aget cljr-magic-require-namespaces short))
-            (list short (list long))))))))
+            (list (list (intern short) (intern long) '()))))))))
```

Comment:
This is ensuring that both the `candidates` path above and the path where we construct candidates from `cljr-magic-require-namespaces` both return a list of triplets, where each triplet contains the alias, namespace, and the list of contexts it appeared in.

Reply:
They are interned because the output from `cljr--get-aliases-from-middleware` are all symbols due to being a filter on the `parseedn-read-str` output. It's possible this could be cleaner if we change the type in `cljr--get-aliases-from-middleware`.